initialization/initializer.go

```diff
@@ -0,0 +1,547 @@
+
+type metadata struct {
+	State metadataState
+	Cfg   config.Config
+}
+
+type Initializer struct {
+	cfg    *Config
+	id     []byte
+	dir    string
+	logger Logger
+}
+
+func NewInitializer(cfg *Config, id []byte) *Initializer {
+	dir := shared.GetInitDir(cfg.DataDir, id)
+	return &Initializer{cfg, id, dir, shared.DisabledLogger{}}
+}
+
+func (init *Initializer) SetLogger(logger Logger) {
+	init.logger = logger
+}
+
+// Initialize takes an id (public key), space (in bytes), numProvenLabels and difficulty.
+// Difficulty determines the number of bits per label that are stored. Each leaf in the tree is 32 bytes = 256 bits.
+// The number of bits per label is 256 / LabelsPerGroup. LabelsPerGroup = 1 << difficulty.
+// Supported values range from 5 (8 bits per label) to 8 (1 bit per label).
+func (init *Initializer) Initialize() (*proving.Proof, error) {
+	if err := ValidateSpace(init.cfg.SpacePerUnit); err != nil {
+		return nil, err
+	}
+
+	difficulty := Difficulty(init.cfg.Difficulty)
+	if err := difficulty.Validate(); err != nil {
+		return nil, err
+	}
+
+	state, requiredSpace, err := init.State()
+	if err != nil {
+		return nil, err
+	}
+
+	if state == StateCompleted {
+		return nil, shared.ErrInitCompleted
+	}
+
+	if !init.cfg.DisableSpaceAvailabilityChecks {
+		availableSpace := shared.AvailableSpace(init.cfg.DataDir)
+		if requiredSpace > availableSpace {
+			return nil, fmt.Errorf("not enough disk space. required: %v, available: %v",
+				bytefmt.ByteSize(requiredSpace), bytefmt.ByteSize(availableSpace))
+		}
+	}
+
+	if err := init.SaveMetadata(MetadataStateStarted); err != nil {
+		return nil, err
+	}
+
+	numFiles, err := NumFiles(init.cfg.SpacePerUnit, init.cfg.FileSize)
+	if err != nil {
+		return nil, err
+	}
+	labelGroupsPerFile := NumLabelGroups(init.cfg.FileSize)
+
+	outputs, err := init.initFiles(numFiles, labelGroupsPerFile)
+	if err != nil {
+		return nil, err
+	}
+
+	output, err := shared.Merge(outputs)
+	if err != nil {
+		return nil, err
+	}
+
+	leafReader := output.Reader.GetLayerReader(0)
+	width, err := leafReader.Width()
+	if err != nil {
+		err = fmt.Errorf("failed to get leaves reader width: %v", err)
+		log.Error(err.Error())
+		return nil, err
+	}
+
+	provenLeafIndices := proving.CalcProvenLeafIndices(output.Root, width<<difficulty, uint8(init.cfg.NumProvenLabels), difficulty)
+	_, provenLeaves, proofNodes, err := merkle.GenerateProof(provenLeafIndices, output.Reader)
+	if err != nil {
+		return nil, err
+	}
+
+	err = leafReader.Close()
+	if err != nil {
+		return nil, err
+	}
+
+	if err := init.SaveMetadata(MetadataStateCompleted); err != nil {
+		return nil, err
+	}
+
+	proof := &proving.Proof{
+		Challenge:    shared.ZeroChallenge,
+		Identity:     init.id,
+		MerkleRoot:   output.Root,
+		ProvenLeaves: provenLeaves,
+		ProofNodes:   proofNodes,
+	}
+
+	return proof, err
+}
+
+func (init *Initializer) Reset() error {
+	if err := init.VerifyStarted(); err != nil {
+		return err
+	}
+
+	err := os.RemoveAll(init.dir)
```

Comment:
Why wouldn't we want to delete the folder with all its files, including `key.bin`?
I think it leaves room for less errors.

Yes, if `Reset` is called, and then `Initialize`, the `key.bin` file won't be added again. But that's because it's not part of the `Initializer` functionality, but rather the init-factory cmd util. We can consider supporting it internally instead.

Reply:
Hmm... On the one hand -- you make a good point. On the other -- while `key.bin` can be created by the init factory, it can now also be created by the node.

I'm not entirely sure when `Reset` is supposed to be called, other than the init factory. If the node calls it and knows to re-create the `key.bin` file it's fine, otherwise we may have an issue.

This dilemma makes me think that using the PoST data directory for the node's key file may have been a bad choice. In any case, we can't just ignore the issue.

We should either clean-up selectively in the PoST or fix it on the node side. Otherwise, if `Reset` shouldn't be called outside of the init factory, then we should remove it from the RPC server.